### PR TITLE
feat(templates): Implement unified lattice module and tests

### PR DESCRIPTION
**Pull request by Stellogic**
Hi @yutuer21,

Following your guidance and the OSPP project requirements ("Object-Oriented Lattice Generation"), I've prepared this initial pull request to add the unified lattice module. I would be grateful for your review and feedback.

**Main Changes in this PR:**

* **`AbstractLattice`:** An abstract base class is proposed to define the common API.
* **Concrete Implementations:** `CustomizeLattice` and `TILattice` (with `SquareLattice`, `HoneycombLattice`, `TriangularLattice` examples) are provided as initial implementations.
* **Core Functionality:** Basic methods like `get_site_info()`, `get_neighbors()`, and `.show()` visualization are included.
* **Testing:** A corresponding test suite (`tests/test_lattice.py`) has been added to cover the core functions and edge cases, achieving ~94% coverage on `lattice.py`.
* **Documentation:** `CHANGELOG.md` has been updated.

This is the first version based on my understanding of the requirements, and I'm sure there is still much room for improvement. Thank you very much for your valuable time and guidance. I look forward to your feedback!

### CODE REVIEW

File: tensorcircuit/templates/lattice.py [hunk 1]
```diff
@@ -0,0 +1,886 @@
...
+    def __len__(self) -> int:
+        """Returns the total number of sites, enabling `len(lattice)`."""
+        return self.num_sites
+
+    # --- Public API for Accessing Lattice Information ---
+
+    def get_coordinates(self, index: SiteIndex) -> Coordinates:
+        """Gets the spatial coordinates of a site by its integer index.
+
+        Args:
```
Comment (refraction-ray) [inline tensorcircuit/templates/lattice.py:78]: for the format of the docstring, please refer to other files in this repo, e.g., https://github.com/tensorcircuit/tensorcircuit-ng/blob/master/tensorcircuit/quantum.py#L1327-L1359

File: tensorcircuit/templates/lattice.py [hunk 2]
```diff
@@ -0,0 +1,886 @@
...
+    Tuple,
+    Union,
+    TYPE_CHECKING,
+)
+
+import matplotlib.pyplot as plt
+
+import numpy as np
+
+if TYPE_CHECKING:
```
Comment (refraction-ray) [inline tensorcircuit/templates/lattice.py:22]: what mypy error is this used for?

File: tensorcircuit/templates/lattice.py [hunk 3]
```diff
@@ -0,0 +1,886 @@
...
+                `matplotlib.pyplot.scatter` function for customizing site appearance.
+        """
+        # creat "fig_created_internally" as flag
+        fig_created_internally = False
+
+        if self.num_sites == 0:
+            print("Lattice is empty, nothing to show.")
+            return
+        if self.dimensionality not in [1, 2, 3]:
+            print(
```
Comment (refraction-ray) [inline tensorcircuit/templates/lattice.py:299]: better directly use ``warning`` or ``logger.warning`` instead of print

**Comment by refraction-ray** [comment]
some more TILattice, 
1D: chain, chain with AB sublattice (ABABAB...), 
2D: rectangular,  checkboard (i.e. square lattice with AB sublattice), Lieb lattice, Kagome lattice 
3D: cubic lattice

**Comment by refraction-ray** [comment]
please also modify __init__.py in templates module accordingly

**Comment by Stellogic** [comment]
Hi @refraction-ray,

Thank you again for your invaluable guidance and detailed review. I have now pushed the updates, incorporating all of your feedback.

Here is a summary of the main changes:

*   **Logging & On-Demand Computation**: Replaced all `print` calls with `logging` and implemented on-demand neighbor calculations.
*   **New `TILattice` Classes**: Added and tested all the suggested lattice implementations (`Chain`, `Kagome`, `Lieb`, etc.).
*   **Code Style & Typing**:
    *   Removed type hints from test function arguments as requested.
    *   In response to your question about the `mypy` error, I have added a detailed comment to the source code explaining the purpose of the `if TYPE_CHECKING:` block for handling the optional `matplotlib` dependency.
*   **Docstring Format Correction**: I have carefully revised all docstrings in `lattice.py` to match the project's required  style.

File: tensorcircuit/templates/__init__.py [hunk 4]
```diff
@@ -7,3 +7,16 @@
 from . import conversions
 
 costfunctions = measurements
+from .lattice import AbstractLattice
```
Comment (refraction-ray) [inline tensorcircuit/templates/__init__.py:10]: just `from . import lattice` as above

File: tensorcircuit/templates/lattice.py [hunk 5]
```diff
@@ -0,0 +1,1198 @@
...
+            comparisons. Defaults to 1e-6.
+        :type tol: float, optional
+        """
+        self._neighbor_maps = {k: {} for k in range(1, max_k + 1)}
+        if self.num_sites < 2:
+            return
+
+        potential_neighbors: Dict[int, List[Tuple[float, int]]] = {}
+        relative_cell_shifts = (
+            np.array([p for p in np.ndindex(*((3,) * self.dimensionality))]) - 1
```
Comment (refraction-ray) [inline tensorcircuit/templates/lattice.py:641]: how can the radius 3 here is sufficient for large `max_k`? the search radius should be propotional to `max_k`?

File: tensorcircuit/templates/lattice.py [hunk 6]
```diff
@@ -0,0 +1,1198 @@
+# -*- coding: utf-8 -*-
+"""
+The lattice module for defining and manipulating lattice geometries.
+"""
+import logging
+
```
Comment (refraction-ray) [inline tensorcircuit/templates/lattice.py:6]: no empty lines here

File: tensorcircuit/templates/lattice.py [hunk 7]
```diff
@@ -0,0 +1,1198 @@
...
+            np.array([p for p in np.ndindex(*((3,) * self.dimensionality))]) - 1
+        )
+
+        for i in range(self.num_sites):
+            # Use a temporary set for each site to automatically handle duplicates
+            neighbors_set: set[Tuple[float, int]] = set()
+
+            ident_i = self._identifiers[i]
+            coord_i = self._coordinates[i]
+            assert isinstance(ident_i, tuple)
```
Comment (refraction-ray) [inline tensorcircuit/templates/lattice.py:650]: no need to assert on variables of the class, such attrs binding should be checked when initialization the class instance if necessary.

File: tests/test_lattice.py [hunk 8]
```diff
@@ -0,0 +1,883 @@
...
+        lattice_4d = CustomizeLattice(
+            dimensionality=4, identifiers=[0], coordinates=[[0, 0, 0, 0]]
+        )
+
+        # Act
+        lattice_4d.show()
+
+        # Assert: Check that the appropriate warning was printed to stdout.
+        assert "show() is not implemented for 4D lattices." in caplog.text
+
```
Comment (refraction-ray) [inline tests/test_lattice.py:850]: I suggest to add one or few tests on relative large lattice, say $6 \times 8$, and to check whether neighbor searching is correctly handled for large `max_k`, say 4, 6, 8 and more, what if the given `max_k` is larger than the possible `max_k`?

**Comment by refraction-ray** [comment]
Thanks for the nice contribution. In general, the code looks very good.

The addition should be a more robust and error-free neighbor searching implementation with careful consideration on PBC and large `max_k`

current failed case for neighbor search:
```python
sq = lattice.SquareLattice(size=(4,4), pbc=True)
len(sq.get_neighbor_pairs(k=3))
# 0
```

File: tensorcircuit/templates/lattice.py [hunk 9]
```diff
@@ -0,0 +1,1403 @@
...
+            identifier next to each site. Defaults to False.
+        :type show_identifiers: bool, optional
+        :param show_bonds_k: Specifies which order of
+            neighbor bonds to draw (e.g., 1 for NN, 2 for NNN). If None,
+            no bonds are drawn. If the specified neighbors have not been
+            calculated, a warning is printed. Defaults to 1.
+        :type show_bonds_k: Optional[int], optional
+        :param ax: An existing Matplotlib Axes object to plot on.
+            If None, a new Figure and Axes are created automatically. Defaults to None.
+        :type ax: Optional["matplotlib.axes.Axes"], optional
```
Comment (refraction-ray) [inline tensorcircuit/templates/lattice.py:302]: no doc for `bond_kwards` arg

Comment (Stellogic) [reply tensorcircuit/templates/lattice.py:302]: Got it, thanks for the review! I'll add the docstring now.

**Comment by Stellogic** [comment]
Hi @refraction-ray 
Thanks for the great suggestions! I've now pushed all the updates.
I've implemented the dynamic modification APIs (from_lattice, add_sites, remove_sites) for CustomizeLattice.
I've fixed the default show() behavior to avoid warnings.

I ran a performance benchmark for the neighbor finding methods. The data shows that the KDTree-based method is likely consistently faster. Based on this, I've kept the implementation simple, using KDTree by default without needing a user-facing switch. The benchmark script is included in templates/ .benchmarks/.

New tests for all new features have been added.

File: tensorcircuit/templates/lattice.py [hunk 10]
```diff
@@ -0,0 +1,1539 @@
...
+                )
+
+                # Calculate squared distances for all images
+                image_d_sq = np.sum(image_displacements**2, axis=2)
+
+                # The true distance is the minimum over all images
+                dist_matrix_sq[i, :] = np.min(image_d_sq, axis=1)
+
+            # Part C: Identify neighbor shells from the correctly calculated distance matrix.
+            # This logic is copied from the working `_find_neighbors_by_distance` method.
```
Comment (refraction-ray) [inline tensorcircuit/templates/lattice.py:707]: `self._distance_matrix` assignment for TILattice mixed boundary condition case

File: tensorcircuit/templates/lattice.py [hunk 11]
```diff
@@ -0,0 +1,1539 @@
...
+                "For better performance, please install it using 'pip install scipy'."
+            )
+            # Fallback to the original, robust-but-slower method
+            return self._build_neighbors_by_distance_matrix(max_k=max_k, tol=tol)
+
+        all_coords = np.array(self._coordinates)
+
+        # 1. Use pdist for memory-efficient calculation of pairwise distances
+        #    to robustly identify the distance shells.
+        all_distances_sq = pdist(all_coords, metric="sqeuclidean")
```
Comment (refraction-ray) [inline tensorcircuit/templates/lattice.py:1370]: `self._distance_matrix` assignment for CustomizeLattice KDTree engine case

**Comment by refraction-ray** [comment]
1. I have added comments for positions to add `self._distance_matrix` assignment for different neighbor building functions, still, the `TILattice` fully PBC case requires several further lines for generating the matrix when build neighbors as the all-to-all distance is not explicitly generated in that case.
2. scipy is always assumed installed as required by `setup.py`, so you can safely delete the check and fallback logic for whether scipy is installed
3. you can move the benchmark script as one example scripts in `/examples` folders

after addressing these points, I believe this PR is perfect to be merged. later you can open new PRs for more examples, tutorials, docs, or feature enhancements.

File: tensorcircuit/templates/.benchmarks/benchmark_neighbors.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+logging.getLogger("tensorcircuit.templates.lattice").setLevel(logging.WARNING)
+
+
+def run_benchmark() -> None:
+    """
+    Executes the benchmark test and prints the results in a formatted table.
+    """
+    site_counts: List[int] = [10, 50, 100, 200, 500, 1000, 1500, 2000]
+    max_k: int = 1
+    number_of_runs: int = 5
```
Comment (refraction-ray) [inline tensorcircuit/templates/.benchmarks/benchmark_neighbors.py:27]: better reduce to 3 for the sake of the total running time of the script

File: examples/lattice_neighbor_benchmark.py [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+To run this script from the project's root directory:
+  python examples/templates/lattice_neighbor_benchmark.py
+"""
+import logging
+import timeit
+from typing import Any, Dict, List
+
+# Suppress verbose logging from the lattice module during the benchmark
+logging.getLogger("tensorcircuit.templates.lattice").setLevel(logging.WARNING)
```
Comment (refraction-ray) [inline examples/lattice_neighbor_benchmark.py:18]: the info level log is ignored by default, I guess removing this line is fine?

File: examples/lattice_neighbor_benchmark.py [hunk 14]
```diff
@@ -0,0 +1,102 @@
+# examples/templates/lattice_neighbor_benchmark.py
```
Comment (refraction-ray) [inline examples/lattice_neighbor_benchmark.py:1]: remove this line

File: tensorcircuit/templates/lattice.py [hunk 15]
```diff
@@ -0,0 +1,1555 @@
...
+                    if is_in_bounds:
+                        target_ident = tuple(target_uc) + (target_basis,)
+                        # The identifier must exist in the map
+                        if target_ident in self._ident_to_idx:
+                            neighbor_idx = self._ident_to_idx[target_ident]
+                            neighbors_of_i_for_k.append(neighbor_idx)
+
+                if neighbors_of_i_for_k:
+                    self._neighbor_maps[k][i] = sorted(neighbors_of_i_for_k)
+        logger.info("Caching the full distance matrix with Minimum Image Convention...")
```
Comment (refraction-ray) [inline tensorcircuit/templates/lattice.py:831]: this `distance_matrix` should be obtained more cleverly, otherwise it will ruin the advantage for an independent neighbor search logic for fully PBC cases, as you again recompute the full distance_matrix from the scratch

File: tensorcircuit/templates/lattice.py (hunk 11 above)
Comment (refraction-ray) [reply tensorcircuit/templates/lattice.py:1379]: again `distance_matrix` should be assembled from `all_distances_sq`, otherwise, the all-to-all distance is computed twice?

**Comment by Stellogic** [comment]
@refraction-ray

Thanks for the review! I've addressed all the feedback and squashed the commits.

File: tensorcircuit/templates/lattice.py (hunk 11 above)
Comment (refraction-ray) [reply tensorcircuit/templates/lattice.py:1379]: distance matrix should also be assigned here, so that after neighbor building, you dont need to recompute `pdist` by calling `_compute_distance_matrix`. The logic, lattice class without neighbor building, distance matrix can be accessed by `_compute_distance_matrix`, lattice class after neighbor building can directly return the matrix without recompute via `pdist` in `_compute_distance_matrix`

File: tests/test_lattice.py [hunk 16]
```diff
@@ -0,0 +1,1276 @@
...
+    def test_modification_clears_distance_matrix_cache(self, initial_lattice):
+        """
+        Tests that add_sites and remove_sites correctly invalidate the
+        cached distance matrix.
+        """
+        # Arrange 1: Compute neighbors and distance matrix
+        lat = initial_lattice
+        lat._build_neighbors(max_k=1)
+        _ = lat.distance_matrix
```
Comment (refraction-ray) [inline tests/test_lattice.py:1185]: a meaningfull test on the matrix instead of a plain assignment?

**Comment by refraction-ray** [comment]
The only remaining part I believe is the improvement on the logic for `distance_matrix` caching when building neighbors. For `CustomizeLattice` case, a simply assignment during neighbor searching could achieve this. For fully PBC case of `TILattice`, the assignment is still missing, hopefully, we can reconstruct the distance matrix based on all the infos from neighbor searching without the direct computation of all-to-all distance matrix. The current independent implementation design for fully PBC case is trying to avoid the full matrix computation after all.

File: tensorcircuit/templates/lattice.py [hunk 17]
```diff
@@ -0,0 +1,1526 @@
...
+        :type max_k: int
+        :param tol: The numerical tolerance to consider two distances equal.
+        :type tol: float
+        :return: A sorted list of squared distances representing the shells.
+        :rtype: List[float]
+        """
+        if not isinstance(all_distances_sq, np.ndarray):
+            all_distances_sq = np.array(all_distances_sq)
+
+        sorted_dist_sq = np.sort(all_distances_sq[all_distances_sq > 1e-12])
```
Comment (refraction-ray) [inline tensorcircuit/templates/lattice.py:480]: why not tol**2 but an explicit value, no need to change the code for this function?

File: tensorcircuit/templates/lattice.py [hunk 18]
```diff
@@ -0,0 +1,1526 @@
...
+            )
+            for source_basis_idx, target_site_idx in zip(*match_indices):
+                source_ident = self._identifiers[source_basis_idx]
+                target_ident = self._identifiers[target_site_idx]
+                uc_disp = np.array(target_ident[:-1]) - np.array(source_ident[:-1])
+                target_basis = target_ident[-1]
+                neighbor_templates[source_basis_idx][k].append((uc_disp, target_basis))
+
+        # 4. Apply templates and SIMULTANEOUSLY reconstruct the distance matrix.
+        self._neighbor_maps = {k: {} for k in range(1, max_k + 1)}
```
Comment (refraction-ray) [inline tensorcircuit/templates/lattice.py:770]: why changing this? the original `self._neighbor_maps = {k: {} for k in range(1, len(dist_shells_sq) + 1)}` is better? which could also free L803-L809, or am I wrong?

File: tensorcircuit/templates/lattice.py [hunk 19]
```diff
@@ -0,0 +1,1526 @@
...
+        if not dist_shells_sq:
+            logger.info("No distinct neighbor shells found.")
+            return
+
+        tree = KDTree(all_coords)
+        self._neighbor_maps = {}
+
+        found_neighbors_so_far: List[set[int]] = [set() for _ in range(self.num_sites)]
+
+        for k_idx, target_d_sq in enumerate(dist_shells_sq):
```
Comment (refraction-ray) [inline tensorcircuit/templates/lattice.py:1347]: why the neigbor building logic here is also modified? The original implementation is more neat? what is the gain here?

File: tests/test_lattice.py [hunk 20]
```diff
@@ -0,0 +1,1540 @@
...
+        # Assert
+        # We compare directly against a pre-sorted list, not a set.
+        # This will fail if the implementation returns [3, 0] instead of [0, 3].
+        assert neighbors == [
+            0,
+            3,
+        ], "The neighbor list should be sorted in ascending order."
+
+
+class TestDistanceMatrixStrength:
```
Comment (refraction-ray) [inline tests/test_lattice.py:1390]: the name convention with Strength is weird, why not only TestDistanceMatrix or something alike?

**Comment by Stellogic** [comment]
@refraction-ray 
My apologies. My process before updating a PR has been to let an AI review it first for any potential issues. I would then follow its suggestions to modify things like variable names, logic, and docstrings. Going forward, I will be more careful in reviewing the AI's suggestions to ensure they are reasonable. Again, I am very sorry."、

**Comment by Stellogic** [comment]
@refraction-ray
My apologies. My process before updating a PR has been to let an AI review it first for any potential issues. I would then follow its suggestions to modify things like variable names, logic, and docstrings. Going forward, I will be more careful in reviewing the AI's suggestions to ensure they are reasonable. Again, I am very sorry.

**Comment by refraction-ray** [comment]
@Stellogic , No need to apologize. It is super okay to use AI, actually I strongly recommend you guys to intensively use AI for coding and development tasks. Just pay attention to the points I mention, as a large bulk of unnecessary changes will make the code review more difficult. AI can write the code for the most parts and everyone use it. It is a good thing, but you should be the first responsible reviewer for what AI generates and changes :)

**Comment by Stellogic** [comment]
Hi @refraction-ray,

Following your guidance, I have force-pushed an update. I've reset the branch back to commit 9ef578d and built a single, clean commit on top of it containing only the necessary fixes to address the test failures.

This new commit includes three targeted fixes:

Fixed Distance Matrix Caching for TILattice: In the fully periodic case, _build_neighbors now reconstructs and caches the full distance matrix by leveraging the existing ref_dist_matrix_sq and translational symmetry. This resolves all failures in TestDistanceMatrix.

Fixed Neighbor Finding for Duplicate Coordinates in 
CustomizeLattice: The logic in _build_neighbors has been refined. It now correctly handles co-located sites by ensuring all points at distance ~0 are excluded from the initial k=1 neighbor shell, which fixes the test_lattice_with_duplicate_coordinates failure.

Improved Numerical Stability in _identify_distance_shells: The logic for filtering zero-distance points has been decoupled from the tolerance-based shell identification. It now robustly filters out self-loops using a fixed small threshold (1e-12), resolving the test_neighbor_shells_with_tiny_separation failure.

File: tensorcircuit/templates/lattice.py [hunk 21]
```diff
@@ -747,6 +744,7 @@ def _build_neighbors(self, max_k: int = 2, **kwargs: Any) -> None:
                     current_k_map[i].sort()
 
                 self._neighbor_maps[k] = current_k_map
+                self._distance_matrix = dist_matrix
```
Comment (refraction-ray) [inline tensorcircuit/templates/lattice.py:747]: wrong tab? why not moving this line directly below L717?

File: tensorcircuit/templates/lattice.py [hunk 22]
```diff
@@ -479,18 +479,15 @@ def _identify_distance_shells(
         if all_distances_sq.size == 0:
             return []
 
-        # Sort unique distances and filter out zero-distance (self-loops)
-        unique_sorted_dist = sorted(
-            [d for d in np.unique(all_distances_sq) if d > tol**2]
-        )
+        sorted_dist = np.sort(all_distances_sq[all_distances_sq > 1e-12])
```
[the remark this answers is not in the document]
Comment (Stellogic) [reply tensorcircuit/templates/lattice.py:482]: I've looked into this, and it seems that a direct replacement with tol**2 causes the test_neighbor_shells_with_tiny_separation test to fail.
 The root cause seemms that tol is treated as a linear distance tolerance, but using tol**2 in the squared-distance comparison logic makes the shell-merging behavior overly sensitive and breaks the test's intent.

**Comment by refraction-ray** [comment]
the CI error seems to be irrelevant for this PR. However, I am still confused on the success of `test_neighbor_shells_with_tiny_separation` function

File: tensorcircuit/templates/lattice.py [hunk 23]
```diff
@@ -0,0 +1,1603 @@
...
+        for i in range(self.num_sites):
+            ident_i = cast(Tuple[Any, ...], self._identifiers[i])
+            uc_i = np.array(ident_i[:-1])
+            basis_i = ident_i[-1]
+
+            uc_disp = -uc_i
+            relative_ucs_list = [
+                cast(Tuple[Any, ...], ident)[:-1] for ident in self._identifiers
+            ]
+            relative_ucs_arr = np.array(relative_ucs_list)
```
Comment (refraction-ray) [inline tensorcircuit/templates/lattice.py:864]: should this be outside the for as it doesn't depend on i?

File: tensorcircuit/templates/lattice.py [hunk 24]
```diff
@@ -0,0 +1,1603 @@
...
+                source_ident = self._identifiers[source_basis_idx]  # (0,..,0,basis)
+                target_ident = self._identifiers[target_site_idx]
+
+                # Rule: (uc_displacement, target_basis_index)
+                uc_disp = np.array(target_ident[:-1]) - np.array(source_ident[:-1])
+                target_basis = target_ident[-1]
+                neighbor_templates[source_basis_idx][k].append((uc_disp, target_basis))
+
+        # 4. Apply the templates to all sites in the lattice
+        self._neighbor_maps = {k: {} for k in range(1, len(dist_shells_sq) + 1)}
```
[the remark this answers is not in the document]
Comment (refraction-ray) [reply tensorcircuit/templates/lattice.py:819]: currently both pbc and obc in TILattice automatically build the whole neighbor maps with invalid `max_k` or `precompute_neighbors`

**Comment by refraction-ray** [comment]
image
my suggestion is to not use the special implementation of fully PBC case, as it is much slower and greatly complicate the implementation

**Comment by refraction-ray** [comment]
the mixed boundary condition universal implementation is efficient enough, just deleting the special tailored implementation branch for fully PBC case, and everything would be nicer and easier

**Comment by refraction-ray** [comment]
@all-contributors please add @Stellogic for code, examples, tests

**Comment by refraction-ray** [comment]
Apart from some more tutorials and examples, the next key milestone could be a machine learning backend agnostic implementation for lattice module, compared to the numpy backend implementation now. A backend agnostic implementation using tensorcircuit.backend API would enable automatic differentiation, vectorized map etc. for this lattice module. A typical example is structure optimization, where the energy computed from the lattice is minimized via gradient descent with tunable parameters the lattice structure parameters.
